DistinctNumeric: Intentionally-incompatible types to type-check IDs and other numerics

BenWiederhake opened the pull request:
This will be mostly useful for the PR I intend to write next: creating a new type `tid_t` and making it incompatible to `pid_t`. There are some weird quirks in the Kernel because these ID spaces aren't properly separated. The idea is to do something like this:

```
typedef /* … */ pid_t;
typedef /* … */ tid_t;

void some_function() {
    pid_t a;
    tid_t b;
    b = a; // Compilation error: Distinct types, even though both are "basically" int!
}
```

Future uses could include coordinate spaces, indexes, units, etc.

Because `DistinctNumeric` is not useful on its own, I intentionally make this a *draft*, because probably the interface needs some changes.

I put this PR up for discussion to get some early feedback: Right now it's easy to change the design :)

---

(In case you're wondering: The 'distinction' comes from using the macros `__FILE__` and `__LINE__` as template parameters. This way, there are no accidental distinction, and accidental identity is very unlikely and can be detected by a simple lint script.)

alimpfard commented on AK/DistinctNumeric.h:
```diff
@@ -0,0 +1,314 @@
+ * The C++20 `operator<=>` would require, among other things `std::weak_equality`.
+ * Since we do not have that, it cannot be implemented.
+ *
+ * The are many operators that do not work on `int`, so I left them out:
+ * `a[b]`, `*a`, `a->b`, `a.b`, `a->*b`, `a.*b`.
```
Not sure if any of these are even useful to have for _numeric_ types, but you can probably use a little template magic (or `requires`, we have concepts...kinda) like `auto helper(a, b) -> decltype(a.b)`...probably.

BenWiederhake replied:
Interesting! But I intend to stick to "just" numeric types. However, I expect to run into something like `some_vector[my_distinct_numeric]`, so I'll probably have to think hard about that, eventually.

BenWiederhake commented:
Changes since the last push:
- Use struct tags instead of file-line tuple (no false negatives anymore)
- Nicer names
- Try and fail to incorporate a nicer scheme like https://gist.github.com/alimpfard/a3b750e8c3a2f44fb3a2d32038968ddf
